python/ray/workflow/step_executor.py

```diff
@@ -194,17 +216,68 @@ async def _write_step_inputs(wf_storage: workflow_storage.WorkflowStorage,
     await asyncio.gather(*save_tasks)
 
 
+def _validate_checkpoint_options(ret: Union["Workflow", Any]):
+    # The condition where checkpointing is not allowed.
+    if not isinstance(ret, Workflow):
+        return
+    context = workflow_context.get_workflow_step_context()
+    if context is None:
+        checkpoint_context = workflow_context.CheckpointContext()
+    else:
+        checkpoint_context = context.checkpoint_context
+    detached_from_dag = checkpoint_context.detached_from_dag
+    if (not detached_from_dag
+            and checkpoint_context.checkpoint != CheckpointMode.SKIP.value):
+        return
+
+    invalid_step = None
+    for w in ret._iter_workflows_in_dag():
+        if w.data.step_options.checkpoint:
+            if w != ret or detached_from_dag:
+                invalid_step = w
+                break
+    if invalid_step is not None:
+        common_str = (f"Step '{invalid_step.step_id}' requires "
+                      "checkpointing, however this is invalid.")
+        if detached_from_dag:
+            raise ValueError(
+                f"{common_str} It is because the outer step "
+                f"skips checkpointing, and the DAG of "
+                f"the sub-workflow is not preserved in the storage. "
+                f"Without the DAG, we cannot recover "
+                f"'{invalid_step.step_id}' from its checkpoint. "
+                f"NOTE: This rule does not "
+                f"affect the output step of the sub-workflow, "
+                f"because once the output step finished, we can "
+                f"recover from the checkpoint of the output step "
+                f"instead of the parent step.")
```

Comment:
Let's try to return only one kind of error here. I'm worried it's still kind of hard to understand, but here is one possible error message.

`checkpoint=True` was passed to the options for this step, but it cannot be checkpointed because it will not be saved in the workflow DAG. This is because its outer workflow has `checkpoint=False`, which means that we will not save the returned nested workflow.